### PR TITLE
TST: Add test for PD hessian

### LINKED ISSUE #3094: BUG: Hessian inversion check is inadequate
The inverse check at

https://github.com/statsmodels/statsmodels/blob/d2de2f0d91e89ff84856d86c2b2d2aa02e6c0fe5/statsmodels/base/model.py#L445

```python
            try:
                Hinv = np.linalg.inv(-1 * self.hessian(xopt))
            except:
                #might want custom warning ResultsWarning? NumericalWarning?
                from warnings import warn
                warndoc = ('Inverting hessian failed, no bse or '
                           'cov_params available')
                warn(warndoc, RuntimeWarning)
                Hinv = None
```

is inadequate and only wil have issues when eigenvalues are very close to zero.  It doesn't verify that all eigenvalues have the same sign (positive, for `-1 * self.hessian(xopt)`).  A better check is to ensure that the minimum eigenvale of `-1 * self.hessian(xopt)` is sufficiently positive and to remove the `try-catch` block since it should not be hit when using a PD hessian.

### LINKED ISSUE #3098: BUG: Kalman filter fails with 1-D complex Z_mat, T_mat, R_mat
This has been happening silently during test_arma_order_select_ic due to the broad `except` inside model.py.  I found this when looking at the Hinv issue.  The solution seems to be to ensure `P` is fortran contiguous 

```
dcomplex[::1,:] P = asfortranarray(dot(pinv(identity(r**2)-kron(T_mat, T_mat)), 
                                   dot(R_mat,R_mat.T).ravel('F')).reshape(r,r, order='F'))
```

The `ValueError` was occurring during the attempt to compute the hessian, and was caught.

The traceback is 

```
Traceback (most recent call last):
  File "C:\git\statsmodels\statsmodels\base\model.py", line 447, in fit
    H = -1 * self.hessian(xopt)
  File "C:\git\statsmodels\statsmodels\tsa\arima_model.py", line 593, in hessian
    return approx_hess_cs(params, self.loglike, args=(False,))
  File "C:\git\statsmodels\statsmodels\tools\numdiff.py", line 243, in approx_hess_cs
    **kwargs)).imag/2./hess[i, j]
  File "C:\git\statsmodels\statsmodels\tsa\arima_model.py", line 776, in loglike
    return self.loglike_kalman(params, set_sigma2)
  File "C:\git\statsmodels\statsmodels\tsa\arima_model.py", line 786, in loglike_kalman
    return KalmanFilter.loglike(params, self, set_sigma2)
  File "C:\git\statsmodels\statsmodels\tsa\kalmanf\kalmanfilter.py", line 666, in loglike
    R_mat, T_mat)
  File "statsmodels\tsa\kalmanf\kalman_loglike.pyx", line 359, in statsmodels.tsa.kalmanf.kalman_loglike.kalman_loglike_complex (statsmodels/tsa/kalmanf/kalman_loglike.c:6432)
    v, F, loglikelihood = kalman_filter_complex(y,k,p,q,r,nobs,Z_mat,R_mat,T_mat)
  File "statsmodels\tsa\kalmanf\kalman_loglike.pyx", line 229, in statsmodels.tsa.kalmanf.kalman_loglike.kalman_filter_complex (statsmodels/tsa/kalmanf/kalman_loglike.c:4956)
    dot(R_mat,R_mat.T).ravel('F')).reshape(r,r, order='F'))
  File "statsmodels\tsa\kalmanf\stringsource", line 644, in View.MemoryView.memoryview_cwrapper (statsmodels/tsa/kalmanf/kalman_loglike.c:14545)
  File "statsmodels\tsa\kalmanf\stringsource", line 345, in View.MemoryView.memoryview.__cinit__ (statsmodels/tsa/kalmanf/kalman_loglike.c:10820)
ValueError: ndarray is not Fortran contiguous
```

### PR DESCRIPTION
**Pull request by bashtage**
Add test for PD hessian
Raise if not PD before inversion

closes #3098
closes #3094

### CODE REVIEW

File: statsmodels/base/model.py [hunk 1]
```diff
@@ -443,8 +443,13 @@ def fit(self, start_params=None, method='newton', maxiter=100,
             Hinv = np.linalg.inv(-retvals['Hessian']) / nobs
         elif not skip_hessian:
             try:
-                Hinv = np.linalg.inv(-1 * self.hessian(xopt))
-            except:
+                H = -1 * self.hessian(xopt)
+                w, v = np.linalg.eigh(H)
```
Comment (josef-pkt) [inline statsmodels/base/model.py]: I usually prefer, `evals, evecs = np.linalg.eigh...` to remember which is which

File: statsmodels/base/model.py [hunk 2]
```diff
@@ -443,8 +443,13 @@ def fit(self, start_params=None, method='newton', maxiter=100,
...
+                if np.min(w) <= 0:
+                    raise RuntimeError
+                Hinv = v.dot(np.diag(1.0 / w)).dot(v.T)
+                Hinv = np.asfortranarray((Hinv + Hinv.T) / 2.0)
+            except RuntimeError:
```
Comment (josef-pkt) [inline statsmodels/base/model.py]: there is no reason anymore to use a try ... except
The warning and alternative can just go below the `if np.min(w) <= 0` and else

File: statsmodels/base/model.py (hunk 1 above)
Comment (josef-pkt) [inline statsmodels/base/model.py]: If we are already calculating the eigenvalues, then we should keep them.
I think we can add them to mlefit.`mle_retvals`, `retvals` see line 474 below

File: statsmodels/tsa/kalmanf/kalman_loglike.pyx [hunk 3]
```diff
@@ -225,8 +225,8 @@ def kalman_filter_complex(dcomplex[:] y,
         dcomplex[::1,:] alpha = PyArray_ZEROS(2, mshape, cnp.NPY_CDOUBLE, FORTRAN)
         int lda = alpha.strides[1]/sizeof(dcomplex)
         # initial variance
-        dcomplex[::1,:] P = dot(pinv(identity(r**2)-kron(T_mat, T_mat)),
-                            dot(R_mat,R_mat.T).ravel('F')).reshape(r,r, order='F')
+        dcomplex[::1,:] P = asfortranarray(dot(pinv(identity(r**2)-kron(T_mat, T_mat)),
+                                           dot(R_mat,R_mat.T).ravel('F')).reshape(r,r, order='F'))
```
Comment (josef-pkt) [inline statsmodels/tsa/kalmanf/kalman_loglike.pyx]: does this need to go into 0.8?
@bashtage @ChadFulton 
I prefer to put the positive definite check into 0.9

Comment (bashtage) [inline statsmodels/tsa/kalmanf/kalman_loglike.pyx]: It is definitely a bug, and one that is hit in testing

**Comment by josef-pkt** [comment]
backport label for kalman_loglike, 
I'm still not sure we want to change the hessian inversion in 0.8

**Comment by bashtage** [comment]
I think it is pretty harmless and provides the user with more information.

**Comment by josef-pkt** [comment]
merged

@bashtage Thank you


I had forgotten that we might have negative eigenvalues in some cases, #3094.
I still think we need better policy on singular hessian handling, but this is an improvement to the current situation (although maybe a bit slower that the previous version for the positive definite check. But we will need the eigenvalues in any case when we want to handle better the singular and near singular cases.)

**Pull request merged**